{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "import itertools\n",
    "import collections\n",
    "import array\n",
    "import numpy as np\n",
    "import numpy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parsiranje referentnog genome iz .fasta formata\n",
    "def readFASTA(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # ignorisi zaglavlje\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "\n",
    "# Parsiranje read-ova i ocena kvaliteta FASTQ fajla\n",
    "def readFASTQ(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # preskoci naziv\n",
    "            seq = fh.readline().rstrip()  # procitaj sekvencu baza\n",
    "            fh.readline()  # preskoci liniju +\n",
    "            qual = fh.readline().rstrip() # procitaj sekvencu ocena kvaliteta\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities\n",
    "\n",
    "def phred33ToQ(qual):\n",
    "    return ord(qual) - 33\n",
    "\n",
    "# vraca reverse complement\n",
    "def reverseComplement(s):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "    t = ''\n",
    "    for base in s:\n",
    "        t = complement[base] + t\n",
    "    return t\n",
    "\n",
    "seqs, quals = readFASTQ('test.fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref = readFASTA('MT.fa')\n",
    "reverseCompl_ref = reverseComplement(ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "reverseSeqs = [] #lista reverse complement sekvenci read-ova test.fastq fajla\n",
    "for i in range(len(seqs)):\n",
    "    reverseSeqs.append(reverseComplement(seqs[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(reverseSeqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SeedP(l, d):   #l - duzina seed-a, d -> duzina rastojanja izmedju seed-ova\n",
    "    seeds = []\n",
    "    for j in range(len(seqs)):\n",
    "        s = []\n",
    "        for i in range(0,len(seqs[j]),d):\n",
    "            s.append(seqs[j][i:(i+l)])\n",
    "        seeds.append(s)\n",
    "    return seeds\n",
    "\n",
    "seeds = SeedP(20,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(seeds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Reverse complement seed-ova\n",
    "def SeedRC(l, d):   #l - duzina seed-a, d -> interval izmedju seed-ova\n",
    "    ReverseCSeeds = []\n",
    "    for j in range(len(reverseSeqs)):\n",
    "        rs = []\n",
    "        for i in range(0,len(reverseSeqs[j]),d):\n",
    "            rs.append(reverseSeqs[j][i:(i+l)])\n",
    "        ReverseCSeeds.append(rs)\n",
    "    return ReverseCSeeds\n",
    "\n",
    "ReverseSeeds = SeedRC(20,10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(ReverseSeeds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Exact matching - naive algorithm\n",
    "\n",
    "\n",
    "# Vraca listu pojavljivanja (offset od pocetka reference)\n",
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    for i in range(len(t) - len(p) + 1):  \n",
    "        match = True\n",
    "        for j in range(len(p)):  \n",
    "            if t[i+j] != p[j]:  \n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i) \n",
    "    return occurrences\n",
    "\n",
    "# naive exact matching algorithm that is strand-aware\n",
    "#Osim sto trazi sve pojave P u T, trazi i pojavu reverse complement-a od P u T \n",
    "def naiveRC(p, t):\n",
    "    r = reverseComplement(p)\n",
    "    if r == p:\n",
    "        return naive(p,t)\n",
    "    else:\n",
    "        return naive(p,t) + naive(r,t)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Exact matching - Boyer Moore\n",
    "\n",
    "import string\n",
    "\n",
    "def z_array(s):\n",
    "    \"\"\" Use Z algorithm (Gusfield theorem 1.4.1) to preprocess s \"\"\"\n",
    "    assert len(s) > 1\n",
    "    z = [len(s)] + [0] * (len(s)-1)\n",
    "    # Initial comparison of s[1:] with prefix\n",
    "    for i in range(1, len(s)):\n",
    "        if s[i] == s[i-1]:\n",
    "            z[1] += 1\n",
    "        else:\n",
    "            break\n",
    "    r, l = 0, 0\n",
    "    if z[1] > 0:\n",
    "        r, l = z[1], 1\n",
    "    for k in range(2, len(s)):\n",
    "        assert z[k] == 0\n",
    "        if k > r:\n",
    "            # Case 1\n",
    "            for i in range(k, len(s)):\n",
    "                if s[i] == s[i-k]:\n",
    "                    z[k] += 1\n",
    "                else:\n",
    "                    break\n",
    "            r, l = k + z[k] - 1, k\n",
    "        else:\n",
    "            # Case 2\n",
    "            # Calculate length of beta\n",
    "            nbeta = r - k + 1\n",
    "            zkp = z[k - l]\n",
    "            if nbeta > zkp:\n",
    "                # Case 2a: Zkp wins\n",
    "                z[k] = zkp\n",
    "            else:\n",
    "                # Case 2b: Compare characters just past r\n",
    "                nmatch = 0\n",
    "                for i in range(r+1, len(s)):\n",
    "                    if s[i] == s[i - k]:\n",
    "                        nmatch += 1\n",
    "                    else:\n",
    "                        break\n",
    "                l, r = k, r + nmatch\n",
    "                z[k] = r - k + 1\n",
    "    return z\n",
    "\n",
    "\n",
    "def n_array(s):\n",
    "    \"\"\" Compile the N array (Gusfield theorem 2.2.2) from the Z array \"\"\"\n",
    "    return z_array(s[::-1])[::-1]\n",
    "\n",
    "\n",
    "def big_l_prime_array(p, n):\n",
    "    \"\"\" Compile L' array (Gusfield theorem 2.2.2) using p and N array.\n",
    "        L'[i] = largest index j less than n such that N[j] = |P[i:]| \"\"\"\n",
    "    lp = [0] * len(p)\n",
    "    for j in range(len(p)-1):\n",
    "        i = len(p) - n[j]\n",
    "        if i < len(p):\n",
    "            lp[i] = j + 1\n",
    "    return lp\n",
    "\n",
    "\n",
    "def big_l_array(p, lp):\n",
    "    \"\"\" Compile L array (Gusfield theorem 2.2.2) using p and L' array.\n",
    "        L[i] = largest index j less than n such that N[j] >= |P[i:]| \"\"\"\n",
    "    l = [0] * len(p)\n",
    "    l[1] = lp[1]\n",
    "    for i in range(2, len(p)):\n",
    "        l[i] = max(l[i-1], lp[i])\n",
    "    return l\n",
    "\n",
    "\n",
    "def small_l_prime_array(n):\n",
    "    \"\"\" Compile lp' array (Gusfield theorem 2.2.4) using N array. \"\"\"\n",
    "    small_lp = [0] * len(n)\n",
    "    for i in range(len(n)):\n",
    "        if n[i] == i+1:  # prefix matching a suffix\n",
    "            small_lp[len(n)-i-1] = i+1\n",
    "    for i in range(len(n)-2, -1, -1):  # \"smear\" them out to the left\n",
    "        if small_lp[i] == 0:\n",
    "            small_lp[i] = small_lp[i+1]\n",
    "    return small_lp\n",
    "\n",
    "\n",
    "def good_suffix_table(p):\n",
    "    \"\"\" Return tables needed to apply good suffix rule. \"\"\"\n",
    "    n = n_array(p)\n",
    "    lp = big_l_prime_array(p, n)\n",
    "    return lp, big_l_array(p, lp), small_l_prime_array(n)\n",
    "\n",
    "\n",
    "def good_suffix_mismatch(i, big_l_prime, small_l_prime):\n",
    "    \"\"\" Given a mismatch at offset i, and given L/L' and l' arrays,\n",
    "        return amount to shift as determined by good suffix rule. \"\"\"\n",
    "    length = len(big_l_prime)\n",
    "    assert i < length\n",
    "    if i == length - 1:\n",
    "        return 0\n",
    "    i += 1  # i points to leftmost matching position of P\n",
    "    if big_l_prime[i] > 0:\n",
    "        return length - big_l_prime[i]\n",
    "    return length - small_l_prime[i]\n",
    "\n",
    "\n",
    "def good_suffix_match(small_l_prime):\n",
    "    \"\"\" Given a full match of P to T, return amount to shift as\n",
    "        determined by good suffix rule. \"\"\"\n",
    "    return len(small_l_prime) - small_l_prime[1]\n",
    "\n",
    "\n",
    "def dense_bad_char_tab(p, amap):\n",
    "    \"\"\" Given pattern string and list with ordered alphabet characters, create\n",
    "        and return a dense bad character table.  Table is indexed by offset\n",
    "        then by character. \"\"\"\n",
    "    tab = []\n",
    "    nxt = [0] * len(amap)\n",
    "    for i in range(0, len(p)):\n",
    "        c = p[i]\n",
    "        assert c in amap\n",
    "        tab.append(nxt[:])\n",
    "        nxt[amap[c]] = i+1\n",
    "    return tab\n",
    "\n",
    "\n",
    "class BoyerMoore(object):\n",
    "    \"\"\" Encapsulates pattern and associated Boyer-Moore preprocessing. \"\"\"\n",
    "    \n",
    "    def __init__(self, p, alphabet='ACGTN'):\n",
    "        self.p = p\n",
    "        self.alphabet = alphabet\n",
    "        # Create map from alphabet characters to integers\n",
    "        self.amap = {}\n",
    "        for i in range(len(self.alphabet)):\n",
    "            self.amap[self.alphabet[i]] = i\n",
    "        # Make bad character rule table\n",
    "        self.bad_char = dense_bad_char_tab(p, self.amap)\n",
    "        # Create good suffix rule table\n",
    "        _, self.big_l, self.small_l_prime = good_suffix_table(p)\n",
    "    \n",
    "    def bad_character_rule(self, i, c):\n",
    "        \"\"\" Return # skips given by bad character rule at offset i \"\"\"\n",
    "        assert c in self.amap\n",
    "        ci = self.amap[c]\n",
    "        assert i > (self.bad_char[i][ci]-1)\n",
    "        return i - (self.bad_char[i][ci]-1)\n",
    "    \n",
    "    def good_suffix_rule(self, i):\n",
    "        \"\"\" Given a mismatch at offset i, return amount to shift\n",
    "            as determined by (weak) good suffix rule. \"\"\"\n",
    "        length = len(self.big_l)\n",
    "        assert i < length\n",
    "        if i == length - 1:\n",
    "            return 0\n",
    "        i += 1  # i points to leftmost matching position of P\n",
    "        if self.big_l[i] > 0:\n",
    "            return length - self.big_l[i]\n",
    "        return length - self.small_l_prime[i]\n",
    "    \n",
    "    def match_skip(self):\n",
    "        \"\"\" Return amount to shift in case where P matches T \"\"\"\n",
    "        return len(self.small_l_prime) - self.small_l_prime[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def boyer_moore(p, p_bm, t):\n",
    "    \"\"\" Do Boyer-Moore matching \"\"\"\n",
    "    i = 0\n",
    "    occurrences = []\n",
    "    while i < len(t) - len(p) + 1:\n",
    "        shift = 1\n",
    "        mismatched = False\n",
    "        for j in range(len(p)-1, -1, -1):\n",
    "            if p[j] != t[i+j]:\n",
    "                skip_bc = p_bm.bad_character_rule(j, t[i+j])\n",
    "                skip_gs = p_bm.good_suffix_rule(j)\n",
    "                shift = max(shift, skip_bc, skip_gs)\n",
    "                mismatched = True\n",
    "                break\n",
    "        if not mismatched:\n",
    "            occurrences.append(i)\n",
    "            skip_gs = p_bm.match_skip()\n",
    "            shift = max(shift, skip_gs)\n",
    "        i += shift\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seed-ovi i pozicije njihovih exact match-eva u referenci koriscenjem naivnog algoritma\n",
    "\n",
    "def pojava(s):\n",
    "    p = {}\n",
    "    for i in range(len(s)):\n",
    "        for j in range(len(s[i])):\n",
    "            p[s[i][j]] = naive(s[i][j], ref)\n",
    "    return p        \n",
    "\n",
    "pojavljivanje = pojava(seeds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(pojavljivanje)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reverse complement seed-ova i pozicije njihovih exact match-eva u referenci koriscenjem naivnog algoritma\n",
    "\n",
    "def pojavaRC(rs):\n",
    "    RCP = {}\n",
    "    for i in range(len(rs)):\n",
    "        for j in range(len(rs[i])):\n",
    "            RCP[rs[i][j]] = naive(rs[i][j], ref)\n",
    "    return RCP\n",
    "\n",
    "ReverseCPojavljivanje = pojavaRC(ReverseSeeds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(ReverseCPojavljivanje)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seed-ovi i pozicije njihovih exact match-eva u referenci koriscenjem Boyer-Moore-ovog algoritma\n",
    "\n",
    "def pojavaBM(s):\n",
    "    pBM = {}\n",
    "    for i in range(len(s)):\n",
    "        for j in range(len(s[i])):\n",
    "            p = s[i][j]\n",
    "            p_bm = BoyerMoore(p, alphabet='ACGTN')\n",
    "            pBM[p] = boyer_moore(p, p_bm, ref)\n",
    "    return(pBM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reverse complement seed-ova i pozicije njihovih exact match-eva u referenci koriscenjem Boyer-Moore-ovog algoritma\n",
    "\n",
    "def pojavaRCBM(rs):\n",
    "    RCPBM = {}\n",
    "    for i in range(len(rs)):\n",
    "        for j in range(len(rs[i])):\n",
    "            p = rs[i][j]\n",
    "            p_bm = BoyerMoore(p, alphabet='ACGTN')\n",
    "            RCPBM[p] = boyer_moore(p, p_bm, ref)\n",
    "    return(RCPBM) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_empty_keys(d):\n",
    "    l = []\n",
    "    for k in list(d):\n",
    "        if not d[k]:\n",
    "            del d[k]\n",
    "\n",
    "            \n",
    "v = dict(pojavljivanje)\n",
    "remove_empty_keys(v)\n",
    "vr = v.values()\n",
    "\n",
    "flattened_vr = [y for x in vr for y in x]\n",
    "\n",
    "#referentni genom se deli na round(max(flattened_vr)/450) prozora cije tezine su smestene u nizu a\n",
    "#prozor dobija vecu tezinu ako postoji veci broj seed-ova koji su se match-ovali na referencu u tom prozoru\n",
    "a = array.array('i',(0 for i in range(0, round(max(flattened_vr)/450))))\n",
    "\n",
    "def pod(niz):\n",
    "    for i in range(len(flattened_vr)):\n",
    "        for j in range(0, round(max(flattened_vr)/450)):\n",
    "            if (flattened_vr[i] > j*450 and flattened_vr[i] < ((j+1)*450-1)):\n",
    "                niz[j] = niz[j] + 1\n",
    "    return(niz)\n",
    "\n",
    "nn = pod(a)\n",
    "#print(nn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "arr = np.array(a)\n",
    "brr = arr.argsort()[-10:][::-1]\n",
    "\n",
    "#lista odabranih prozora u referenci sortiranih po tezini\n",
    "ref_iz = []\n",
    "for i in range(len(brr)):\n",
    "    ref_iz.append(ref[brr[i]*450:(brr[i]+1)*450])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(ref_iz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "vRC = dict(ReverseCPojavljivanje)\n",
    "remove_empty_keys(vRC)\n",
    "vrRC = vRC.values()\n",
    "#print(vrRC)\n",
    "\n",
    "flattened_vrRC = [y for x in vrRC for y in x]\n",
    "#print(flattened_vrRC)\n",
    "\n",
    "#referentni genom se deli na round(max(flattened_vr)/450) prozora cije tezine su smestene u nizu aRC\n",
    "#prozor dobija vecu tezinu ako postoji veci broj reverse complement seed-ova koji su se match-ovali na referencu u tom prozoru\n",
    "aRC = array.array('i',(0 for i in range(0, round(max(flattened_vrRC)/450))))\n",
    "\n",
    "def podRC(nizRC):\n",
    "    for i in range(len(flattened_vrRC)):\n",
    "        for j in range(0, round(max(flattened_vrRC)/450)):\n",
    "            if (flattened_vrRC[i] > j*450 and flattened_vrRC[i] < ((j+1)*450-1)):\n",
    "                nizRC[j] = nizRC[j] + 1\n",
    "    return(nizRC)\n",
    "\n",
    "nnRC = podRC(aRC)\n",
    "#print(nnRC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "arrRC = np.array(aRC)\n",
    "brrRC = arrRC.argsort()[-10:][::-1]\n",
    "\n",
    "#lista odabranih prozora u referenci sortiranih po tezini\n",
    "ref_izRC = []\n",
    "for i in range(len(brrRC)):\n",
    "    ref_izRC.append(ref[brrRC[i]*450:(brrRC[i]+1)*450])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(ref_izRC)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Local Alignment\n",
    "\n",
    "def scoringMatrix(a,b):\n",
    "    if a==b: return 1 # match\n",
    "    if a=='_' or b=='_': return -7 # indel\n",
    "    return -4 # mismatch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def localAlignment(x,y, s):\n",
    "    D = numpy.zeros((len(x)+1,len(y)+1), dtype=int)\n",
    "\n",
    "    for i in range(1,len(x)+1):\n",
    "        for j in range(1,len(y)+1):\n",
    "            D[i,j]=max(0, \n",
    "                       D[i-1,j]  +s(x[i-1], '_'),\n",
    "                       D[i,j-1]  +s('_',   y[j-1]), \n",
    "                       D[i-1,j-1]+s(x[i-1],y[j-1]))\n",
    "    \n",
    "    # find the cell in the table which has maximum value       \n",
    "    localMax = D.max()\n",
    "    return D, localMax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "x=ref_iz[0]\n",
    "y = seqs[5]\n",
    "V, score = localAlignment(x, y, scoringMatrix)\n",
    "#print(V)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(\"Best score is: %d in cell %s\" % (score, numpy.unravel_index(numpy.argmax(V), V.shape)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tracebak(x,y,V, s):\n",
    "    maxValue=numpy.where(V==V.max())\n",
    "    i=maxValue[0][0]\n",
    "    j=maxValue[1][0]\n",
    "    ax,ay,am, tr = '','','',''\n",
    "    while (i>0 or j>0) and V[i,j]!=0:\n",
    "        if i>0 and j>0:\n",
    "            sigma = 1 if x[i-1]==y[j-1] else 0\n",
    "            d=V[i-1,j-1]+s(x[i-1],y[j-1]) # diagonal move\n",
    "        if i>0: v=V[i-1,j]+s(x[i-1],'_')  # vertical move\n",
    "        if j>0: h=V[i,j-1]+s('_',y[j-1])  # horizontal move\n",
    "        \n",
    "        # diagonal is the best\n",
    "        if d>=v and d>=h:\n",
    "            ax += x[i-1]\n",
    "            ay += y[j-1]\n",
    "            if sigma==1:\n",
    "                tr+='M'\n",
    "                am+='|'\n",
    "            else:\n",
    "                tr+='R'\n",
    "                am+=' '\n",
    "            i-=1\n",
    "            j-=1\n",
    "        # vertical is the best\n",
    "        elif v>=h:\n",
    "            ax+=x[i-1]\n",
    "            ay+='_'\n",
    "            tr+='I'\n",
    "            am+=' '\n",
    "            i-=1\n",
    "        # horizontal is the best\n",
    "        else:\n",
    "            ay+=y[j-1]\n",
    "            ax+='_'\n",
    "            tr+='D'\n",
    "            am+=' '\n",
    "            j-=1\n",
    "    alignment= '\\n'.join([ax[::-1], am[::-1], ay[::-1]])\n",
    "    return alignment, tr[::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SAMoutput(self, outf, read, pos, max, t):\n",
    "        with open(outf) as file:\n",
    "            file.write(\"{}  {}  {}  {}  {}  {}  {}  {}\\n\".format(\n",
    "                read[0], \"MT\", pos, max, t, read[1], read[2]))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
